### PR TITLE
[FEA] New Graph Interface and Loaders for Distributed Sampling in DGL

### LINKED ISSUE #3303: Getter and setter functions of dgl.graph in cugraph storage
Is it possible to add all the getter and setter functions of the dgl.graph into the cugraph storage? For instance, we use the following code;

`compare = graph.nodes["node1"].data.get("data1")`

to get the data of a node.

### LINKED ISSUE #3805: Refactor cuGraph-PyG and cuGraph-DGL Tests to Match cuGraph Standard
cuGraph recently refactored its tests to make MG testing easier, and allow for testing MG workflows with only 1 GPU.  The test filenames and annotations were also simplified.  cuGraph-PyG and cuGraph-DGL need to refactor their tests as well.  A major bug in cuGraph-PyG involving repartitioning of the dask edgelist went unnoticed because the tests were not refactored.  Finally doing this refactor will prevent similar issues in the future.

### LINKED ISSUE #4247: [FEA] New Distributed Graph Store for cuGraph-DGL
This new distributed graph store will not use Dask.  Instead, each worker will pass in a slice of the edgelist, and a `pylibcugraph.MGGraph` will be created on each worker.  This store should also be completely unaware of feature storage.  It will be fully compatible with the distributed sampler (#4246).

### LINKED ISSUE #4400: WholeFeatureStorage for cuGraph-DGL
Support WholeGraph feature storage in cuGraph-PyG through a new `WholeFeatureStore` class that implements the `FeatureStorage` interface.

### PR DESCRIPTION
**Pull request by alexbarghi-nv**
- [x] New Graph class that duck-types `DGLGraph`
- [x] New WholeGraph feature store that implements DGL's `FeatureStorage`
- [x] New Sampler interface
- [x] New Loader interface
- [x] New Sampler for distributed uniform neighbor sampling without dask
- [x] New Loader for distributed uniform neighbor sampling without dask

This PR also adds `pydantic` and `torchdata` as test dependencies of cuGraph-DGL, because they are unlisted runtime dependencies of the latest version of DGL (which we will be testing with in the new `cugraph-gnn` repository).

This PR also changes the test file structure to match the rest of cuGraph, by using the __mg_ suffix to denote multi-GPU tests and moving them to the main directory under "tests".

New examples will be added in a future PR.  This PR is just for basic functionality and tests.

Closes #3805 
Closes rapidsai/cugraph-gnn#4
Closes #4400
Closes #4247
Closes #3303

### CODE REVIEW

File: python/cugraph-dgl/cugraph_dgl/dataloading/dataloader.py
Comment (alexbarghi-nv) [inline python/cugraph-dgl/cugraph_dgl/dataloading/dataloader.py:1]: Sorry about this diff, this is really a new file that took the name from what is now `dask_dataloader.py`

File: python/cugraph-dgl/cugraph_dgl/tests/test_graph_mg.py [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+    torch.distributed.init_process_group("nccl", rank=rank, world_size=world_size)
+
+    pylibwholegraph.torch.initialize.init(
+        rank,
+        world_size,
+        rank,
+        world_size,
+    )
+
+    cugraph_comms_init(rank=rank, world_size=world_size, uid=cugraph_id, device=rank)
```
Comment (VibhuJawa) [inline python/cugraph-dgl/cugraph_dgl/tests/test_graph_mg.py:72]: Do we have a util around this instead ?

**Review by VibhuJawa — COMMENTED** [review]
Mostly looks good to me, Have added some nits. I think we should add a function to go from dgl.DGLGraph to cugraph_dgl.Graph , having that will help a lot in testing and usability for folks.

File: python/cugraph-dgl/cugraph_dgl/dataloading/dataloader.py [hunk 2]
```diff
@@ -164,158 +134,33 @@ def __init__(
...
+        return self.__dataset
 
     def __iter__(self):
-        output_dir = os.path.join(
-            self._sampling_output_dir, "epoch_" + str(self.epoch_number)
+        # TODO move to the correct device
+        return self.__sampler.sample(
+            self.__graph,
+            self.__dataset,
+            batch_size=self.__batch_size,
```
[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply python/cugraph-dgl/cugraph_dgl/dataloading/dataloader.py:157]: The `DistSampler` is constructed here: https://github.com/rapidsai/cugraph/blob/5e511cc77092350602bc8d19fcd2a23306e99089/python/cugraph-dgl/cugraph_dgl/dataloading/neighbor_sampler.py#L183

Sampling happens here: https://github.com/rapidsai/cugraph/blob/b55c279b8865b962a8757d1235d3309938fbb354/python/cugraph/cugraph/gnn/data_loading/dist_sampler.py#L736 and it calls the appropriate function based on whether the graph is a `SGGraph` or `MGGraph`, which is wrapped by the `is_multi_gpu` function.

[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply python/cugraph-dgl/cugraph_dgl/dataloading/dataloader.py:157]: So the short answer is neither; it's using the new `DistSampler` which automatically handles whether or not the graph is SG or MG.

File: python/cugraph-dgl/cugraph_dgl/dataloading/dataloader.py [hunk 3]
```diff
@@ -164,158 +134,33 @@ def __init__(
...
-def _clean_directory(path):
-    """param <path> could either be relative or absolute."""
-    if os.path.isfile(path):
-        os.remove(path)  # remove the file
-    elif os.path.isdir(path):
-        shutil.rmtree(path)  # remove dir and all contains
+        start = start.to(self.__device)
+        end = end.to(self.__device)
+        blocks = [b.to(self.__device) for b in blocks]
+        """
```
[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply python/cugraph-dgl/cugraph_dgl/dataloading/dataloader.py]: No; removed it.

**Comment by alexbarghi-nv** [comment]
> Mostly looks good to me, Have added some nits. I think we should add a function to go from dgl.DGLGraph to cugraph_dgl.Graph , having that will help a lot in testing and usability for folks.

I added `cugraph_dgl_graph_from_heterograph` which converts a DGLGraph to a cugraph_dgl.Graph along with appropriate tests.

File: python/cugraph-dgl/cugraph_dgl/tests/test_graph_mg.py (hunk 1 above)
Comment (alexbarghi-nv) [reply python/cugraph-dgl/cugraph_dgl/tests/test_graph_mg.py]: Moved this to a utility.

**Review by VibhuJawa — APPROVED** [review]
LGTM, thanks for all the changes @alexbarghi-nv , Looks good to go from my side

**Review by raydouglass — APPROVED** [review]

File: python/cugraph-dgl/cugraph_dgl/graph.py [hunk 4]
```diff
@@ -0,0 +1,910 @@
...
+from cugraph_dgl.view import (
+    HeteroNodeView,
+    HeteroNodeDataView,
+    HeteroEdgeView,
+    HeteroEdgeDataView,
+)
+
+
+# Have to use import_optional even though these are required
+# dependencies in order to build properly.
```
Comment (rlratzel) [inline python/cugraph-dgl/cugraph_dgl/graph.py:36]: That's interesting, and unfortunate.  I'm glad you added this comment.

**Review by rlratzel — APPROVED** [review]

**Pull request merged**